### DISCUSSION TITLE
Building on mac os using flag -DLIBUS_USE_LIBUV still gives error

### DISCUSSION THREAD

**Discussion by afterlifenirvana** [uNetworking/uWebSockets#886] [Q&A] [answered]
clang -DLIBUS_NO_SSL -DLIBUS_USE_LIBUV -flto -O3 -c -IuSockets/src uSockets/src/*.c uSockets/src/eventing/*.c
clang++ -DLIBUS_NO_SSL -DUWS_NO_ZLIB -DLIBUS_USE_LIBUV -flto -O3 -c -std=c++17 -Isrc -IuSockets/src examples/HelloWorld.cpp
clang++ -flto -O3 -s *.o -o HelloWorld
ld: warning: option -s is obsolete and being ignored
undef: _uv_poll_stop
undef: _uv_ref
undef: _uv_loop_delete
undef: _uv_is_closing
undef: _uv_loop_new
undef: _uv_prepare_stop
undef: _uv_close
undef: _uv_prepare_init
undef: _uv_prepare_start
undef: _uv_async_init
undef: _uv_unref
undef: _uv_check_stop
undef: _uv_timer_start
undef: _uv_check_init
undef: _uv_timer_stop
undef: _uv_timer_init
undef: _uv_poll_init_socket
undef: _uv_poll_start
undef: _uv_default_loop
undef: _uv_check_start
undef: _uv_run
Undefined symbols for architecture x86_64:
  "_uv_poll_stop", referenced from:
      _us_socket_close in lto.o
  "_uv_ref", referenced from:
      _us_loop_free in lto.o
  "_uv_loop_delete", referenced from:
      _us_loop_free in lto.o
  "_uv_is_closing", referenced from:
      _check_cb in lto.o
  "_uv_loop_new", referenced from:
      _us_create_loop in lto.o
  "_uv_prepare_stop", referenced from:
      _us_loop_free in lto.o
  "_uv_close", referenced from:
      _us_loop_free in lto.o
      _us_socket_close in lto.o
  "_uv_prepare_init", referenced from:
      _us_create_loop in lto.o
  "_uv_prepare_start", referenced from:
      _us_create_loop in lto.o
  "_uv_async_init", referenced from:
      _us_create_loop in lto.o
  "_uv_unref", referenced from:
      _us_create_loop in lto.o
  "_uv_check_stop", referenced from:
      _us_loop_free in lto.o
  "_uv_timer_start", referenced from:
      _main in lto.o
      _us_create_loop in lto.o
  "_uv_check_init", referenced from:
      _us_create_loop in lto.o
  "_uv_timer_stop", referenced from:
      _us_loop_free in lto.o
  "_uv_timer_init", referenced from:
      _us_create_loop in lto.o
  "_uv_poll_init_socket", referenced from:
      _main in lto.o
      _poll_cb in lto.o
  "_uv_poll_start", referenced from:
      _main in lto.o
      uWS::AsyncSocket<false>::write(char const*, int, bool, int) in lto.o
      _poll_cb in lto.o
  "_uv_default_loop", referenced from:
      _us_create_loop in lto.o
  "_uv_check_start", referenced from:
      _us_create_loop in lto.o
  "_uv_run", referenced from:
      _main in lto.o
      _us_loop_free in lto.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [examples] Error 1

**Comment by ghost** [comment]
Please don't send build errors as big reports. You need to learn how to crawl before you try and fly.

**Comment by afterlifenirvana** [comment]
Can you give a hint ? this is the example build file from the repo with no changes. What am i suppose to fix here ?

**Comment by ghost** [comment]
No

**Comment by afterlifenirvana** [comment]
ok :)

**Comment by ghost** [comment]
I can admit the makefile is lacking for anything non-Linux. But the issue you have is such a default issue, posted over and over again by people who haven't learned the basics of their tools yet. I can see if I can add macOS to the makefile but like I said it should be obvious

**Comment by sab24** [comment]
@afterlifenirvana you need to link to libuv

See my answer at:
https://stackoverflow.com/questions/55591983/how-to-fix-fatal-error-sys-epoll-h-file-not-found-in-osx/56057892#56057892

**Comment by afterlifenirvana** [comment]
@sab24  I have used -DLIBUS_USE_LIBUV if you would see traceback you can notice that. I solved the sys/epoll.h after doing some research. I have posted this cause of the linker error which I'm getting even after using  flags.

**Comment by sab24** [comment]
Yeah, you also need -luv to link it to the uv library in the last command

**Comment by ghost** [comment, marked as answer]
What is the point in using an advanced C++ library with a very specific audience if you cannot even figure out the most basic linker issue? I'm sorry for being frank but you need to go back to the drawing board and go over the basics of C++ and Unix before posting issues here. One Google search and you will find a stack overflow post with like 10k upvotes

**Comment by afterlifenirvana** [comment]
I agree you are right. I'm noob at c++.  @sab24  Thank you for pointing out the mistake. @alexhultman you will get developers who will try to use your advanced library even after being noob at c++. I want to use this library even after i haven't learned c++ because it is a great library fit for my use and not having learned c++ won't stop me because i know i can take a look at those 10k upvotes answer and still build stuff but some time we miss things and i guess its a open community so we should help otherwise what's the point of having a community. I understand you must be frustrated after answering same build questions again and again Sorry for the trouble. I will try getting better at it. Anyway i started using it on a linux system the day i posted this question.

**Comment by sab24** [comment]
Yeah I agree, we all have to start somewhere, and without the help from others computers were never built and C++ never existed. I also agree with doing research by yourself first to reduce the load on others, as asking too many questions can be annoying. But in this case due diligence has been showed, and I think we should help those people.
